File: libraries/node/src/sources/sm/convert.ts
/**
 * Mapper from the StepMania source model onto URC charts.
 */
import { UrcError } from "../../error.js";
import type { Chart, NoteType } from "../../model.js";
import { buildTiming, checkHoldOverlap, roundMs } from "../shared.js";
import type { SmChart, SmFile, SmNote, Timing } from "./model.js";
import { resolveLanes } from "./parse.js";
import { filterNotes, preprocess, rows, warpIntervals } from "./preprocess.js";

const MEASURE_ROWS = 192;
const ROLL_TAP_SPACING_MS = 500;

const TYPE_ORDER: Record<NoteType, number> = { N: 0, LS: 1, LE: 2, M: 3, F: 4 };

const DIFFICULTY_NAMES: Record<string, string> = {
	beginner: "Beginner",
	easy: "Easy",
	basic: "Easy",
	light: "Easy",
	medium: "Medium",
	another: "Medium",
	trick: "Medium",
	standard: "Medium",
	difficult: "Medium",
	hard: "Hard",
	ssr: "Hard",
	maniac: "Hard",
	heavy: "Hard",
	smaniac: "Challenge",
	challenge: "Challenge",
	expert: "Challenge",
	oni: "Challenge",
	edit: "Edit"
};

/**
 * Converts every chart of a simfile into URC charts.
 */
export function convertSm(simfile: SmFile): Chart[] {
	return simfile.charts.map(chart => convertChart(simfile, chart));
}

type Entry = [number, number, string, unknown];

function convertChart(simfile: SmFile, chart: SmChart): Chart {
	const lanes = resolveLanes(chart.stepsType);
	const timing = chart.timing ?? simfile.timing;
	const [offset, bpmSegs, stopSegs, warpSegs] = preprocess(timing);
	const intervals = warpIntervals([...warpSegs, ...timing.warps]);
	const notes = filterNotes(chart.notes, intervals);

	const entries: Entry[] = [];
	for (const [start, dest] of intervals)
		entries.push([start, 0, "warp", dest]);
	for (const [beat, seconds] of timing.delays)
		entries.push([rows(beat), 1, "delay", seconds]);
	for (let index = 0; index < notes.length; index++) {
		entries.push([notes[index].row, 2, "note", index]);
		const tailRow = notes[index].tailRow;
		if (tailRow !== null)
			entries.push([tailRow, 2, "tail", index]);
	}
	for (const [beat, seconds] of stopSegs)
		entries.push([rows(beat), 3, "stop", seconds]);
	for (const [beat, bpm] of bpmSegs)
		entries.push([rows(beat), 4, "bpm", bpm]);
	for (const [beat, numerator, denominator] of timing.timesigs)
		entries.push([rows(beat), 5, "timesig", [numerator, denominator] as [number, number]]);
	for (const [beat, ratio] of timing.scrolls)
		entries.push([rows(beat), 6, "scroll", ratio]);
	const maxRow = entries.length > 0 ? Math.max(...entries.map(entry => entry[0])) : 0;
	for (let row = 0; row < maxRow + MEASURE_ROWS; row += MEASURE_ROWS)
		entries.push([row, 2, "anchor", null]);

	const headTimes: number[] = new Array<number>(notes.length).fill(0);
	const tailTimes = new Map<number, number>();
	const bpmPoints: Array<[number, number, number, number]> = [];
	const svPoints: Array<[number, number]> = [];
	const anchors: number[] = [];

	let seconds = -offset;
	let bpm: number | null = null;
	let meter: [number, number] = [4, 4];
	let multiplier = 1;
	let warping = false;
	let warpDest = 0;
	let prevRow: number | null = null;

	entries.sort((left, right) => left[0] - right[0] || left[1] - right[1]);
	for (let i = 0; i < entries.length; ) {
		const row = entries[i][0];
		if (prevRow !== null && !warping && bpm !== null)
			seconds += ((row - prevRow) / 48) * 60 / bpm;
		prevRow = row;
		if (warping && row >= warpDest)
			warping = false;

		let newBpm: number | null = bpm;
		let newMeter: [number, number] = meter;
		let newMultiplier = multiplier;
		for (; i < entries.length && entries[i][0] === row; i++) {
			const [, , kind, value] = entries[i];
			switch (kind) {
				case "warp": {
					const dest = value as number;
					if (warping)
						warpDest = Math.max(warpDest, dest);
					else {
						warping = true;
						warpDest = dest;
					}
					break;
				}
				case "delay":
					seconds += value as number;
					break;
				case "note":
					headTimes[value as number] = seconds;
					break;
				case "tail":
					tailTimes.set(value as number, seconds);
					break;
				case "anchor":
					anchors.push(roundMs(seconds * 1000));
					break;
				case "stop":
					seconds += value as number;
					break;
				case "bpm":
					newBpm = value as number;
					break;
				case "timesig":
					newMeter = value as [number, number];
					break;
				default:
					newMultiplier = value as number;
			}
		}

		if (newBpm !== null && (newBpm !== bpm || newMeter[0] !== meter[0] || newMeter[1] !== meter[1]))
			bpmPoints.push([roundMs(seconds * 1000), newBpm, newMeter[0], newMeter[1]]);
		if (newMultiplier !== multiplier)
			svPoints.push([roundMs(seconds * 1000), newMultiplier]);
		bpm = newBpm;
		meter = newMeter;
		multiplier = newMultiplier;
	}

	const urcNotes = buildUrcNotes(timing, notes, headTimes, tailTimes);
	const noteTimes = urcNotes.filter(([, , type]) => type !== "LE").map(([time]) => time);
	const firstNoteTime = noteTimes.length > 0 ? Math.min(...noteTimes) : 0;

	const timingPoints = buildTiming(
		bpmPoints,
		svPoints,
		firstNoteTime,
		".sm",
		anchors.find(time => time >= firstNoteTime) ?? null
	);

	urcNotes.sort(
		(left, right) => left[0] - right[0] || left[1] - right[1] || TYPE_ORDER[left[2]] - TYPE_ORDER[right[2]]
	);
	const finalNotes = urcNotes.map(([time, lane, type]) => ({
		timestampMs: time - firstNoteTime,
		lane,
		type
	}));
	checkHoldOverlap(finalNotes);

	const title = [simfile.title, simfile.subtitle].filter(part => part !== "").join(" ");
	return {
		formatVersion: { major: 1, minor: 1 },
		metadata: {
			original: "StepMania",
			title: title === "" ? "Unknown" : title,
			artist: simfile.artist === "" ? "Unknown" : simfile.artist,
			creator: chart.credit || simfile.credit || "Unknown",
			version: chart.chartname || difficultyName(chart.difficulty, chart.description)
		},
		judgment: null,
		layout: { keys: lanes, specialKeys: 0, specialLanes: null },
		timing: timingPoints,
		notes: finalNotes
	};
}

type UrcNote = [number, number, NoteType];

function buildUrcNotes(
	timing: Timing,
	notes: SmNote[],
	headTimes: number[],
	tailTimes: Map<number, number>
): UrcNote[] {
	const fakeRanges = timing.fakes.map(
		([beat, length]) => [rows(beat), rows(beat) + rows(length)] as [number, number]
	);
	const urcNotes: UrcNote[] = [];

	for (let index = 0; index < notes.length; index++) {
		const note = notes[index];
		const headMs = roundMs(headTimes[index] * 1000);
		if (fakeRanges.some(([start, end]) => start <= note.row && note.row < end)) {
			urcNotes.push([headMs, note.track, "F"]);
			continue;
		}
		if (note.kind === "hold") {
			const tailMs = roundMs((tailTimes.get(index) ?? 0) * 1000);
			if (tailMs <= headMs)
				throw new UrcError("syntax", 1, `hold on lane ${note.track} collapses to zero length`);
			urcNotes.push([headMs, note.track, "LS"]);
			urcNotes.push([tailMs, note.track, "LE"]);
		} else if (note.kind === "roll") {
			const endMs = roundMs((tailTimes.get(index) ?? 0) * 1000);
			urcNotes.push([headMs, note.track, "N"]);
			for (let tapMs = headMs + ROLL_TAP_SPACING_MS; tapMs < endMs; tapMs += ROLL_TAP_SPACING_MS)
				urcNotes.push([tapMs, note.track, "N"]);
		} else if (note.kind === "mine")
			urcNotes.push([headMs, note.track, "M"]);
		else if (note.kind === "fake")
			urcNotes.push([headMs, note.track, "F"]);
		else
			urcNotes.push([headMs, note.track, "N"]);
	}
	return urcNotes;
}

function difficultyName(difficulty: string, description: string): string {
	let name = DIFFICULTY_NAMES[difficulty.trim().toLowerCase()] ?? "";
	if (name === "Hard" && ["smaniac", "challenge"].includes(description.trim().toLowerCase()))
		name = "Challenge";
	return name === "" ? "Edit" : name;
}

File: libraries/rust/src/sources/sm/convert.rs
//! Mapper from the StepMania source model onto URC charts.

use crate::error::Result;
use crate::model::{Chart, Layout, Metadata, Note, NoteType, Version};

use super::super::shared::{build_timing, check_hold_overlap, round_ms};
use super::model::{SmChart, SmFile};
use super::notes::{build_urc_notes, difficulty_name, first_non_empty};
use super::parse::resolve_lanes;
use super::preprocess::{filter_notes, preprocess, rows, warp_intervals};

const MEASURE_ROWS: i64 = 192;

type BpmPoint = (i64, f64, u64, u64);

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum EntryKind {
    Warp,
    Delay,
    Note,
    Tail,
    Anchor,
    Stop,
    Bpm,
    TimeSignature,
    Scroll,
}

#[derive(Debug, Clone, Copy)]
enum EntryValue {
    Number(f64),
    NoteIndex(usize),
    Meter(u64, u64),
}

/// Converts every chart of a simfile into URC charts.
pub fn convert_sm(simfile: &SmFile) -> Result<Vec<Chart>> {
    simfile.charts.iter().map(|chart| convert_chart(simfile, chart)).collect()
}

fn convert_chart(simfile: &SmFile, chart: &SmChart) -> Result<Chart> {
    let lanes = resolve_lanes(&chart.steps_type)?;
    let timing = chart.timing.as_ref().unwrap_or(&simfile.timing);
    let (offset, bpm_segs, stop_segs, warp_segs) = preprocess(timing)?;
    let mut all_warps = warp_segs;
    all_warps.extend_from_slice(&timing.warps);
    let intervals = warp_intervals(&all_warps);
    let mut notes = chart.notes.clone();
    filter_notes(&mut notes, &intervals);

    let mut entries: Vec<(i64, u8, EntryKind, EntryValue)> = Vec::new();
    for &(start, dest) in &intervals {
        entries.push((start, 0, EntryKind::Warp, EntryValue::Number(dest as f64)));
    }
    for &(beat, seconds) in &timing.delays {
        entries.push((rows(beat), 1, EntryKind::Delay, EntryValue::Number(seconds)));
    }
    for (index, note) in notes.iter().enumerate() {
        entries.push((note.row, 2, EntryKind::Note, EntryValue::NoteIndex(index)));
        if let Some(tail_row) = note.tail_row {
            entries.push((tail_row, 2, EntryKind::Tail, EntryValue::NoteIndex(index)));
        }
    }
    for &(beat, seconds) in &stop_segs {
        entries.push((rows(beat), 3, EntryKind::Stop, EntryValue::Number(seconds)));
    }
    for &(beat, bpm) in &bpm_segs {
        entries.push((rows(beat), 4, EntryKind::Bpm, EntryValue::Number(bpm)));
    }
    for &(beat, numerator, denominator) in &timing.timesigs {
        entries.push((
            rows(beat),
            5,
            EntryKind::TimeSignature,
            EntryValue::Meter(numerator, denominator),
        ));
    }
    for &(beat, ratio) in &timing.scrolls {
        entries.push((rows(beat), 6, EntryKind::Scroll, EntryValue::Number(ratio)));
    }
    let max_row = entries.iter().map(|entry| entry.0).max().unwrap_or(0);
    let mut row = 0;
    while row < max_row + MEASURE_ROWS {
        entries.push((row, 2, EntryKind::Anchor, EntryValue::Number(0.0)));
        row += MEASURE_ROWS;
    }

    let mut head_times = vec![0.0_f64; notes.len()];
    let mut tail_times = vec![0.0_f64; notes.len()];
    let mut bpm_points: Vec<BpmPoint> = Vec::new();
    let mut sv_points: Vec<(i64, f64)> = Vec::new();
    let mut anchors: Vec<i64> = Vec::new();

    let mut seconds = -offset;
    let mut bpm: Option<f64> = None;
    let mut meter = (4_u64, 4_u64);
    let mut multiplier = 1.0_f64;
    let mut warping = false;
    let mut warp_dest = 0_i64;
    let mut prev_row: Option<i64> = None;

    entries.sort_by_key(|entry| (entry.0, entry.1));
    let mut i = 0;
    while i < entries.len() {
        let row = entries[i].0;
        if let Some(prev) = prev_row
            && !warping
            && let Some(current) = bpm
        {
            seconds += (row - prev) as f64 / 48.0 * 60.0 / current;
        }
        prev_row = Some(row);
        if warping && row >= warp_dest {
            warping = false;
        }

        let mut new_bpm = bpm;
        let mut new_meter = meter;
        let mut new_multiplier = multiplier;
        while i < entries.len() && entries[i].0 == row {
            let (_, _, kind, value) = &entries[i];
            match (kind, value) {
                (EntryKind::Warp, EntryValue::Number(dest)) => {
                    let dest = *dest as i64;
                    if warping {
                        warp_dest = warp_dest.max(dest);
                    } else {
                        warping = true;
                        warp_dest = dest;
                    }
                }
                (EntryKind::Delay, EntryValue::Number(pause)) => seconds += *pause,
                (EntryKind::Note, EntryValue::NoteIndex(index)) => {
                    head_times[*index] = seconds;
                }
                (EntryKind::Tail, EntryValue::NoteIndex(index)) => {
                    tail_times[*index] = seconds;
                }
                (EntryKind::Anchor, _) => anchors.push(round_ms(seconds * 1000.0)),
                (EntryKind::Stop, EntryValue::Number(pause)) => seconds += *pause,
                (EntryKind::Bpm, EntryValue::Number(value)) => new_bpm = Some(*value),
                (EntryKind::TimeSignature, EntryValue::Meter(num, den)) => {
                    new_meter = (*num, *den);
                }
                (EntryKind::Scroll, EntryValue::Number(ratio)) => new_multiplier = *ratio,
                _ => unreachable!("entry kind/value mismatch"),
            }
            i += 1;
        }

        if let Some(emitted_bpm) = new_bpm
            && (new_bpm != bpm || new_meter != meter)
        {
            bpm_points.push((round_ms(seconds * 1000.0), emitted_bpm, new_meter.0, new_meter.1));
        }
        if new_multiplier != multiplier {
            sv_points.push((round_ms(seconds * 1000.0), new_multiplier));
        }
        bpm = new_bpm;
        meter = new_meter;
        multiplier = new_multiplier;
    }

    let urc_notes = build_urc_notes(timing, &notes, &head_times, &tail_times)?;
    let first_note_time = urc_notes
        .iter()
        .filter(|note| note.2 != NoteType::Le)
        .map(|note| note.0)
        .min()
        .unwrap_or(0);

    let timing_points = build_timing(
        &bpm_points,
        &sv_points,
        first_note_time,
        ".sm",
        anchors.iter().copied().find(|time| *time >= first_note_time),
    )?;

    let type_order = |kind: NoteType| match kind {
        NoteType::N => 0,
        NoteType::Ls => 1,
        NoteType::Le => 2,
        NoteType::M => 3,
        NoteType::F => 4,
    };
    let mut ordered: Vec<(i64, u32, NoteType)> = urc_notes;
    ordered.sort_by_key(|note| (note.0, note.1, type_order(note.2)));
    let final_notes: Vec<Note> = ordered
        .into_iter()
        .map(|(time, lane, note_type)| Note {
            timestamp_ms: time - first_note_time,
            lane,
            note_type,
        })
        .collect();
    check_hold_overlap(&final_notes)?;

    let title = [simfile.title.as_str(), simfile.subtitle.as_str()]
        .iter()
        .filter(|part| !part.is_empty())
        .cloned()
        .collect::<Vec<&str>>()
        .join(" ");
    Ok(Chart {
        format_version: Version { major: 1, minor: 1 },
        metadata: Metadata {
            original: "StepMania".to_string(),
            title: if title.is_empty() { "Unknown".to_string() } else { title },
            artist: if simfile.artist.is_empty() {
                "Unknown".to_string()
            } else {
                simfile.artist.clone()
            },
            creator: first_non_empty(&chart.credit, &simfile.credit, "Unknown").to_string(),
            version: if chart.chartname.is_empty() {
                difficulty_name(&chart.difficulty, &chart.description).to_string()
            } else {
                chart.chartname.clone()
            },
        },
        judgment: None,
        layout: Layout {
            keys: lanes as u64,
            special_keys: 0,
            special_lanes: None,
        },
        timing: timing_points,
        notes: final_notes,
    })
}


